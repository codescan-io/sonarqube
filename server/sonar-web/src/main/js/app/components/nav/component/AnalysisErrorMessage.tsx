/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { Link } from 'design-system';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import { useLocation } from 'react-router-dom';
import { hasMessage, translate } from '../../../../helpers/l10n';
import { getComponentBackgroundTaskUrl } from '../../../../helpers/urls';
import { Task } from '../../../../types/tasks';
import { Component } from '../../../../types/types';

interface Props {
  component: Component;
  currentTask: Task;
  currentTaskOnSameBranch?: boolean;
  onLeave: () => void;
}

export function AnalysisErrorMessage(props: Props) {
  const { component, currentTask, currentTaskOnSameBranch } = props;

  const location = useLocation();

  const backgroundTaskUrl = getComponentBackgroundTaskUrl(component.key);
  const canSeeBackgroundTasks = component.configuration?.showBackgroundTasks;
  const isOnBackgroundTaskPage = location.pathname === backgroundTaskUrl.pathname;

  const branch =
    currentTask.branch ??
    `${currentTask.pullRequest ?? ''}${
      currentTask.pullRequestTitle ? ' - ' + currentTask.pullRequestTitle : ''
    }`;

  let messageKey;
  if (currentTaskOnSameBranch === false && branch) {
    messageKey = 'component_navigation.status.failed_branch';
  } else {
    messageKey = 'component_navigation.status.failed';
  }

  let type;
  if (hasMessage('background_task.type', currentTask.type)) {
    messageKey += '_X';
    type = translate('background_task.type', currentTask.type);
  }

  let url;
  let stacktrace;
  if (canSeeBackgroundTasks) {
    messageKey += '.admin';

    if (isOnBackgroundTaskPage) {
      messageKey += '.help';
      stacktrace = translate('background_tasks.show_stacktrace');
    } else {
      messageKey += '.link';
      url = (
        <Link onClick={props.onLeave} to={backgroundTaskUrl}>
          {translate('background_tasks.page')}
        </Link>
      );
    }
  }

  return (
    <FormattedMessage
      defaultMessage={translate(messageKey)}
      id={messageKey}
      values={{ branch, url, stacktrace, type }}
    />
  );
}
