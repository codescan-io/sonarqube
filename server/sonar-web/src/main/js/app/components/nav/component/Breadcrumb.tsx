/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { last } from 'lodash';
import * as React from 'react';
import Link from '../../../../components/common/Link';
import QualifierIcon from '../../../../components/icons/QualifierIcon';
import { isMainBranch } from '../../../../helpers/branch-like';
import { getComponentOverviewUrl } from '../../../../helpers/urls';
import { BranchLike } from '../../../../types/branch-like';
import { Component } from '../../../../types/types';
import { colors } from '../../../theme';

export interface BreadcrumbProps {
  component: Component;
  currentBranchLike: BranchLike | undefined;
}

export function Breadcrumb(props: BreadcrumbProps) {
  const {
    component: { breadcrumbs },
    currentBranchLike,
  } = props;
  const lastBreadcrumbElement = last(breadcrumbs);
  const isNotMainBranch = currentBranchLike && !isMainBranch(currentBranchLike);

  return (
    <div className="big flex-shrink display-flex-center">
      {breadcrumbs.map((breadcrumbElement, i) => {
        const isFirst = i === 0;
        const isNotLast = i < breadcrumbs.length - 1;
        const isLast = !isNotLast;
        const showQualifierIcon = isFirst && lastBreadcrumbElement;

        const name =
          isNotMainBranch || isNotLast ? (
            <>
              {showQualifierIcon && !isNotMainBranch && (
                <QualifierIcon
                  className="spacer-right"
                  qualifier={lastBreadcrumbElement.qualifier}
                  fill={colors.neutral800}
                />
              )}
              <Link
                className="link-no-underline"
                to={getComponentOverviewUrl(breadcrumbElement.key, breadcrumbElement.qualifier)}
              >
                {showQualifierIcon && isNotMainBranch && (
                  <QualifierIcon
                    className="spacer-right"
                    qualifier={lastBreadcrumbElement.qualifier}
                    fill={colors.primary}
                  />
                )}
                {breadcrumbElement.name}
              </Link>
            </>
          ) : (
            <>
              {showQualifierIcon && (
                <QualifierIcon
                  className="spacer-right"
                  qualifier={lastBreadcrumbElement.qualifier}
                  fill={colors.neutral800}
                />
              )}
              {breadcrumbElement.name}
            </>
          );

        return (
          <span className="flex-shrink display-flex-center" key={breadcrumbElement.key}>
            {isLast ? (
              <h1 className="text-ellipsis" title={breadcrumbElement.name}>
                {name}
              </h1>
            ) : (
              <span className="text-ellipsis" title={breadcrumbElement.name}>
                {name}
              </span>
            )}
            {isNotLast && <span className="slash-separator" />}
          </span>
        );
      })}
    </div>
  );
}

export default React.memo(Breadcrumb);
