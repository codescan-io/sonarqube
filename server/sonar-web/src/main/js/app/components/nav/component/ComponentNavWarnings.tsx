/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import AnalysisWarningsModal from '../../../../components/common/AnalysisWarningsModal';
import { Alert } from '../../../../components/ui/Alert';
import { translate } from '../../../../helpers/l10n';
import { TaskWarning } from '../../../../types/tasks';

interface Props {
  componentKey: string;
  isBranch: boolean;
  onWarningDismiss: () => void;
  warnings: TaskWarning[];
}

interface State {
  modal: boolean;
}

export default class ComponentNavWarnings extends React.PureComponent<Props, State> {
  state: State = { modal: false };

  handleClick = (event: React.MouseEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    this.setState({ modal: true });
  };

  handleCloseModal = () => {
    this.setState({ modal: false });
  };

  render() {
    return (
      <>
        <Alert className="js-component-analysis-warnings flex-1" display="inline" variant="warning">
          <FormattedMessage
            defaultMessage={translate('component_navigation.last_analysis_had_warnings')}
            id="component_navigation.last_analysis_had_warnings"
            values={{
              branchType: this.props.isBranch
                ? translate('branches.branch')
                : translate('branches.pr'),
              warnings: (
                <a href="#" onClick={this.handleClick}>
                  <FormattedMessage
                    defaultMessage={translate('component_navigation.x_warnings')}
                    id="component_navigation.x_warnings"
                    values={{
                      warningsCount: this.props.warnings.length,
                    }}
                  />
                </a>
              ),
            }}
          />
        </Alert>
        {this.state.modal && (
          <AnalysisWarningsModal
            componentKey={this.props.componentKey}
            onClose={this.handleCloseModal}
            onWarningDismiss={this.props.onWarningDismiss}
            warnings={this.props.warnings}
          />
        )}
      </>
    );
  }
}
