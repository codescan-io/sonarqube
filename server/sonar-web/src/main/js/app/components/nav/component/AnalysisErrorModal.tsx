/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { ButtonLink } from '../../../../components/controls/buttons';
import Modal from '../../../../components/controls/Modal';
import { hasMessage, translate } from '../../../../helpers/l10n';
import { Task } from '../../../../types/tasks';
import { Component } from '../../../../types/types';
import { AnalysisErrorMessage } from './AnalysisErrorMessage';
import { AnalysisLicenseError } from './AnalysisLicenseError';

interface Props {
  component: Component;
  currentTask: Task;
  currentTaskOnSameBranch?: boolean;
  onClose: () => void;
}

export function AnalysisErrorModal(props: Props) {
  const { component, currentTask, currentTaskOnSameBranch } = props;

  const header = translate('error');

  const licenseError =
    currentTask.errorType &&
    hasMessage('license.component_navigation.button', currentTask.errorType);

  return (
    <Modal contentLabel={header} onRequestClose={props.onClose}>
      <header className="modal-head">
        <h2>{header}</h2>
      </header>

      <div className="modal-body modal-container">
        {licenseError ? (
          <AnalysisLicenseError currentTask={currentTask} />
        ) : (
          <AnalysisErrorMessage
            component={component}
            currentTask={currentTask}
            currentTaskOnSameBranch={currentTaskOnSameBranch}
            onLeave={props.onClose}
          />
        )}
      </div>

      <footer className="modal-foot">
        <ButtonLink className="js-modal-close" onClick={props.onClose}>
          {translate('close')}
        </ButtonLink>
      </footer>
    </Modal>
  );
}
