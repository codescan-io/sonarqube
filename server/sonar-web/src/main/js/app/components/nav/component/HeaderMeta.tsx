/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { useIntl } from 'react-intl';
import BranchStatus from '../../../../components/common/BranchStatus';
import Link from '../../../../components/common/Link';
import HomePageSelect from '../../../../components/controls/HomePageSelect';
import { formatterOption } from '../../../../components/intl/DateTimeFormatter';
import { isBranch, isPullRequest } from '../../../../helpers/branch-like';
import { translate, translateWithParameters } from '../../../../helpers/l10n';
import { BranchLike } from '../../../../types/branch-like';
import { ComponentQualifier } from '../../../../types/component';
import { TaskWarning } from '../../../../types/tasks';
import { Component } from '../../../../types/types';
import { CurrentUser, HomePage, isLoggedIn } from '../../../../types/users';
import withCurrentUserContext from '../../current-user/withCurrentUserContext';
import ComponentNavWarnings from './ComponentNavWarnings';
import './HeaderMeta.css';

export interface HeaderMetaProps {
  branchLike?: BranchLike;
  currentUser: CurrentUser;
  component: Component;
  onWarningDismiss: () => void;
  warnings: TaskWarning[];
}

export function HeaderMeta(props: HeaderMetaProps) {
  const { branchLike, component, currentUser, warnings } = props;

  const isABranch = isBranch(branchLike);

  const currentPage = getCurrentPage(component, branchLike);
  const displayVersion = component.version !== undefined && isABranch;
  const lastAnalysisDate = useIntl().formatDate(component.analysisDate, formatterOption);

  return (
    <>
      <div className="display-flex-center flex-0 small">
        {warnings.length > 0 && (
          <span className="header-meta-warnings">
            <ComponentNavWarnings
              isBranch={isABranch}
              componentKey={component.key}
              onWarningDismiss={props.onWarningDismiss}
              warnings={warnings}
            />
          </span>
        )}
        {component.analysisDate && (
          <span
            title={translateWithParameters(
              'overview.project.last_analysis.date_time',
              lastAnalysisDate
            )}
            className="spacer-left nowrap note"
          >
            {lastAnalysisDate}
          </span>
        )}
        {displayVersion && (
          <span className="spacer-left nowrap note">{`${translate('version')} ${
            component.version
          }`}</span>
        )}
        {isLoggedIn(currentUser) && currentPage !== undefined && !isPullRequest(branchLike) && (
          <HomePageSelect className="spacer-left" currentPage={currentPage} />
        )}
      </div>
      {isPullRequest(branchLike) && (
        <div className="navbar-context-meta-secondary display-inline-flex-center">
          {branchLike.url !== undefined && (
            <Link
              className="link-no-underline big-spacer-right"
              to={branchLike.url}
              target="_blank"
              size={12}
            >
              {translate('branches.see_the_pr')}
            </Link>
          )}
          <BranchStatus branchLike={branchLike} component={component} />
        </div>
      )}
    </>
  );
}

export function getCurrentPage(component: Component, branchLike: BranchLike | undefined) {
  let currentPage: HomePage | undefined;

  const branch = isBranch(branchLike) && !branchLike.isMain ? branchLike.name : undefined;

  switch (component.qualifier) {
    case ComponentQualifier.Portfolio:
    case ComponentQualifier.SubPortfolio:
      currentPage = { type: 'PORTFOLIO', component: component.key };
      break;
    case ComponentQualifier.Application:
      currentPage = {
        type: 'APPLICATION',
        component: component.key,
        branch,
      };
      break;
    case ComponentQualifier.Project:
      // when home page is set to the default branch of a project, its name is returned as `undefined`
      currentPage = {
        type: 'PROJECT',
        component: component.key,
        branch,
      };
      break;
  }

  return currentPage;
}

export default withCurrentUserContext(HeaderMeta);
