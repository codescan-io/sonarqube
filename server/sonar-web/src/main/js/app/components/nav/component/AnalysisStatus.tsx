/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { DeferredSpinner, FlagMessage, Link } from 'design-system';
import * as React from 'react';
import AnalysisWarningsModal from '../../../../components/common/AnalysisWarningsModal';
import { translate } from '../../../../helpers/l10n';
import { Task, TaskStatuses, TaskWarning } from '../../../../types/tasks';
import { Component } from '../../../../types/types';
import { AnalysisErrorModal } from './AnalysisErrorModal';

export interface HeaderMetaProps {
  currentTask?: Task;
  currentTaskOnSameBranch?: boolean;
  component: Component;
  isInProgress?: boolean;
  isPending?: boolean;
  onWarningDismiss: () => void;
  warnings: TaskWarning[];
}

export function AnalysisStatus(props: HeaderMetaProps) {
  const { component, currentTask, currentTaskOnSameBranch, isInProgress, isPending, warnings } =
    props;

  const [modalIsVisible, setDisplayModal] = React.useState(false);
  const openModal = React.useCallback(() => {
    setDisplayModal(true);
  }, [setDisplayModal]);
  const closeModal = React.useCallback(() => {
    setDisplayModal(false);
  }, [setDisplayModal]);

  if (isInProgress || isPending) {
    return (
      <div className="sw-flex sw-items-center">
        <DeferredSpinner timeout={0} />
        <span className="sw-ml-1">
          {isInProgress
            ? translate('project_navigation.analysis_status.in_progress')
            : translate('project_navigation.analysis_status.pending')}
        </span>
      </div>
    );
  }

  if (currentTask?.status === TaskStatuses.Failed) {
    return (
      <>
        <FlagMessage ariaLabel={translate('alert.tooltip.error')} variant="error">
          <span>{translate('project_navigation.analysis_status.failed')}</span>
          <Link
            className="sw-ml-1"
            blurAfterClick={true}
            onClick={openModal}
            preventDefault={true}
            to={{}}
          >
            {translate('project_navigation.analysis_status.details_link')}
          </Link>
        </FlagMessage>
        {modalIsVisible && (
          <AnalysisErrorModal
            component={component}
            currentTask={currentTask}
            currentTaskOnSameBranch={currentTaskOnSameBranch}
            onClose={closeModal}
          />
        )}
      </>
    );
  }

  if (warnings.length > 0) {
    return (
      <>
        <FlagMessage ariaLabel={translate('alert.tooltip.warning')} variant="warning">
          <span>{translate('project_navigation.analysis_status.warnings')}</span>
          <Link
            className="sw-ml-1"
            blurAfterClick={true}
            onClick={openModal}
            preventDefault={true}
            to={{}}
          >
            {translate('project_navigation.analysis_status.details_link')}
          </Link>
        </FlagMessage>
        {modalIsVisible && (
          <AnalysisWarningsModal
            componentKey={component.key}
            onClose={closeModal}
            taskId={currentTask?.id}
            onWarningDismiss={props.onWarningDismiss}
            warnings={warnings}
          />
        )}
      </>
    );
  }

  return null;
}
