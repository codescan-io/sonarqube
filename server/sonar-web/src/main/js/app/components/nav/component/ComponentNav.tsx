/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { LAYOUT_GLOBAL_NAV_HEIGHT, LAYOUT_PROJECT_NAV_HEIGHT, TopBar } from 'design-system';
import * as React from 'react';
import { translate } from '../../../../helpers/l10n';
import {
  ProjectAlmBindingConfigurationErrors,
  ProjectAlmBindingResponse,
} from '../../../../types/alm-settings';
import { BranchLike } from '../../../../types/branch-like';
import { ComponentQualifier } from '../../../../types/component';
import { Task, TaskWarning } from '../../../../types/tasks';
import { Component, Organization } from '../../../../types/types';
import RecentHistory from '../../RecentHistory';
import ComponentNavProjectBindingErrorNotif from './ComponentNavProjectBindingErrorNotif';
import Header from './Header';
import HeaderMeta from './HeaderMeta';
import Menu from './Menu';
import InfoDrawer from './projectInformation/InfoDrawer';
import ProjectInformation from './projectInformation/ProjectInformation';

export interface ComponentNavProps {
  branchLikes: BranchLike[];
  currentBranchLike: BranchLike | undefined;
  component: Component;
  currentTask?: Task;
  currentTaskOnSameBranch?: boolean;
  isInProgress?: boolean;
  isPending?: boolean;
  onComponentChange: (changes: Partial<Component>) => void;
  onWarningDismiss: () => void;
  projectBinding?: ProjectAlmBindingResponse;
  projectBindingErrors?: ProjectAlmBindingConfigurationErrors;
  warnings: TaskWarning[];
  organization: Organization;
  comparisonBranchesEnabled: boolean;
}

export default function ComponentNav(props: ComponentNavProps) {
  const {
    branchLikes,
    component,
    currentBranchLike,
    currentTask,
    currentTaskOnSameBranch,
    isInProgress,
    isPending,
    projectBinding,
    projectBindingErrors,
    warnings,
    organization,
    comparisonBranchesEnabled,
  } = props;

  const [displayProjectInfo, setDisplayProjectInfo] = React.useState(false);

  React.useEffect(() => {
    const { breadcrumbs, key, name } = component;
    const { qualifier } = breadcrumbs[breadcrumbs.length - 1];
    if (
      [
        ComponentQualifier.Project,
        ComponentQualifier.Portfolio,
        ComponentQualifier.Application,
        ComponentQualifier.Developper,
      ].includes(qualifier as ComponentQualifier)
    ) {
      RecentHistory.add(key, name, qualifier.toLowerCase());
    }
  }, [component, component.key]);

  let prDecoNotifComponent;
  if (projectBindingErrors !== undefined) {
    prDecoNotifComponent = <ComponentNavProjectBindingErrorNotif component={component} />;
  }

  return (
    <>
      <TopBar id="context-navigation" aria-label={translate('qualifier', component.qualifier)}>
        <div className="sw-flex sw-justify-between">
          <Header
            branchLikes={branchLikes}
            component={component}
            currentBranchLike={currentBranchLike}
            projectBinding={projectBinding}
            organization={organization}
            comparisonBranchesEnabled={comparisonBranchesEnabled}
          />
          <HeaderMeta
            branchLike={currentBranchLike}
            component={component}
            currentTask={currentTask}
            currentTaskOnSameBranch={currentTaskOnSameBranch}
            isInProgress={isInProgress}
            isPending={isPending}
            onWarningDismiss={props.onWarningDismiss}
            warnings={warnings}
          />
        </div>
        <Menu
          branchLike={currentBranchLike}
          branchLikes={branchLikes}
          component={component}
          isInProgress={isInProgress}
          isPending={isPending}
          onToggleProjectInfo={() => {
            setDisplayProjectInfo(!displayProjectInfo);
          }}
          projectInfoDisplayed={displayProjectInfo}
          comparisonBranchesEnabled={comparisonBranchesEnabled}
        />
        <InfoDrawer
          displayed={displayProjectInfo}
          onClose={() => {
            setDisplayProjectInfo(false);
          }}
          top={LAYOUT_GLOBAL_NAV_HEIGHT + LAYOUT_PROJECT_NAV_HEIGHT}
        >
          <ProjectInformation
            branchLike={currentBranchLike}
            component={component}
            onComponentChange={props.onComponentChange}
          />
        </InfoDrawer>
      </TopBar>
      {prDecoNotifComponent}
    </>
  );
}
