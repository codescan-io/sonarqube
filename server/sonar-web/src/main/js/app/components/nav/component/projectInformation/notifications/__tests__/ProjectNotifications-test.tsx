/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { mockComponent } from '../../../../../../../helpers/mocks/component';
import {
  NotificationGlobalType,
  NotificationProjectType,
} from '../../../../../../../types/notifications';
import { ProjectNotifications } from '../ProjectNotifications';

jest.mock('react', () => {
  return {
    ...jest.requireActual('react'),
    useEffect: jest.fn().mockImplementation((f) => f()),
    useRef: jest.fn().mockReturnValue({ current: document.createElement('h3') }),
  };
});

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('should add and remove a notification for the project', () => {
  const addNotification = jest.fn();
  const removeNotification = jest.fn();
  const wrapper = shallowRender({ addNotification, removeNotification });
  const notification = {
    channel: 'EmailNotificationChannel',
    type: 'SQ-MyNewIssues',
  };

  wrapper.find('NotificationsList').prop<Function>('onAdd')(notification);
  expect(addNotification).toHaveBeenCalledWith({ ...notification, project: 'foo' });

  wrapper.find('NotificationsList').prop<Function>('onRemove')(notification);
  expect(removeNotification).toHaveBeenCalledWith({ ...notification, project: 'foo' });
});

it('should set focus on the heading when rendered', () => {
  const fakeElement = document.createElement('h3');
  const focus = jest.fn();
  (React.useRef as jest.Mock).mockReturnValueOnce({ current: { ...fakeElement, focus } });

  shallowRender();
  expect(focus).toHaveBeenCalled();
});

function shallowRender(props = {}) {
  return shallow(
    <ProjectNotifications
      addNotification={jest.fn()}
      channels={['channel1', 'channel2']}
      component={mockComponent({ key: 'foo' })}
      globalTypes={[NotificationGlobalType.CeReportTaskFailure, NotificationGlobalType.NewAlerts]}
      loading={false}
      notifications={[
        {
          channel: 'channel1',
          type: 'type-global',
          project: 'foo',
          projectName: 'Foo',
        },
        {
          channel: 'channel1',
          type: 'type-common',
          project: 'bar',
          projectName: 'Bar',
        },
        {
          channel: 'channel2',
          type: 'type-common',
          project: 'qux',
          projectName: 'Qux',
        },
      ]}
      perProjectTypes={[NotificationProjectType.CeReportTaskFailure]}
      removeNotification={jest.fn()}
      {...props}
    />
  );
}
