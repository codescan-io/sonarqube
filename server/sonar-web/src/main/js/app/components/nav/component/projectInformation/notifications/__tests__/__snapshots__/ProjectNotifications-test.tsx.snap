// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Fragment>
  <h3
    tabIndex={-1}
  >
    project.info.notifications
  </h3>
  <Alert
    aria-live="off"
    className="spacer-top"
    variant="info"
  >
    notification.dispatcher.information
  </Alert>
  <DeferredSpinner
    loading={false}
  >
    <table
      className="data zebra notifications-table"
    >
      <thead>
        <tr>
          <th
            aria-label="project"
          />
          <th
            className="text-center"
            key="channel1"
          >
            <h4>
              notification.channel.channel1
            </h4>
          </th>
          <th
            className="text-center"
            key="channel2"
          >
            <h4>
              notification.channel.channel2
            </h4>
          </th>
        </tr>
      </thead>
      <NotificationsList
        channels={
          [
            "channel1",
            "channel2",
          ]
        }
        checkboxId={[Function]}
        notifications={
          [
            {
              "channel": "channel1",
              "project": "foo",
              "projectName": "Foo",
              "type": "type-global",
            },
          ]
        }
        onAdd={[Function]}
        onRemove={[Function]}
        project={true}
        types={
          [
            "CeReportTaskFailure",
          ]
        }
      />
    </table>
  </DeferredSpinner>
</Fragment>
`;
